src/sentry/sentry_metrics/consumers/indexer/batch.py

```diff
@@ -79,6 +101,32 @@ def extract_strings(self) -> Mapping[int, Set[str]]:
                 )
                 continue
 
+    @metrics.wraps("process_messages.limit_messages")
+    def limit_messages(self) -> None:
```

Comment:
Any reason we cannot follow the same pattern we have in the processing method today ?
https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/consumers/indexer/processing.py#L43-L56

Basically making the `IndexerBatch` just a data structure that tweaks our batch (transform, filter, etc).
As we removed the indexer dependency from the batch to keep the single responsibility we should be able to move the rate limiter outside. 

I think a way to achieve it is to proceed this way:
- Allow the `IndexerBatch` to return the list of time series contained in the batch (either as hash or as a more complex data structure). Maybe the hash is enough.
- The `reconstruct_message` is already capable of filtering the content of the batch based on this `Mapping[int, Mapping[str, Metadata]]` which is the `bulk_record_metadata` argument. I think we should have a cleaner way to filter the `IndexerBatch`. Imagine: `filter(timeseries: Set[Hash])` or `filter(messages: Set[Offset])`.
- Now we would have the following: `IndexerBatch` is capable of batching messages, reconstructing messages, returning aggregated views of the batch (strings, hashes) or filter it.  So any logic that needs to do something of application relevance with the batch can safely stay out of the batch and the batch only tweaks data structures.
- With this the process could be:  1) create batch, 2) extract hashes, 3) dimensionality limter admission control, 4) filter batch, 5) write on indexer, 6) reconstruct.


What do you think ?
My reason for this is to keep as much as possible out of `IdnexerBatch` which is already annoyingly complex. So responsibilities that are not directly tied to transforming the batch are the first we could move out.

Reply:
you're right. i moved the extraction into the message processor. there are still areas where the PR still significantly differs from your proposed solution which I think are worth discussing:

* The message payload is passed all the way into `limiters.cardinality`, instead of being converted into a hash in IndexerBatch. The idea is that the rate limiter can do an early return if there are no quotas to be applied, without computing hashes, or do other similar optimizations.
* The access to the message content is typed out using `InboundMessage` to contain just the fields that the rate limiter is supposed to access. Casting the message to a typeddict with a subset of defined keys instead of creating a separate dataclass/namedtuple is also cheaper at runtime.

let me know if that is okay... basically `IndexerBatch` is still a pure datastructure, but instead of doing rate limiter specific things like hashing it is just leaking the entire message payload now